### ISSUE TITLE
Does Burrow support Consumer Group based Lag calculation configuration

### ISSUE THREAD

**Issue by aravindan-bk** [linkedin/Burrow#27]
Hi,
  I have a use-case where we have setup Kafka for 2 Topics. Each topic has consumers who consume messages at a different rate. For example Consumer_Topic_1 might do a lot of work after consuming a message and before committing the Offset. Consumer_Topic_2 might do half of work done by Consumer_Topic_1. 
 
 In this case, their sense of Slowness (or Lag) are very different. Hence ideally, I would need 2 sets of [ticker] and [lagcheck] sections, each serving a Consumer Group. But, I believe today it's not possible in a single instance of Burrow. Is this understanding correct?

 I probably need to deploy 2 instances of Burrow to achieve what I need. How viable is this ask for a new feature to be added? Are there any plans around this line?

Thanks in advance!

**Comment by toddpalino** [comment]
I'm not sure I understand the problem that you're running into. The only configuration under ticker is how often the brokers are queried for the current (tail) offsets for each partition. This is independent of any consumer groups and is used to calculate lag as consumer offsets come in. The configurations under lagcheck determine how many offset commits for a consumer group are stored and evaluated on, and how long to wait before expiring an inactive group.

Burrow already does all calculation specific to the consumer group. The lag and status is calculated for each group independently. How often the consumer group commits only affects the time period that the group is evaluated over. For example, let's say that group A commits offsets every 60 seconds, and group B commits offsets every 10 minutes. If the "intervals" parameter is set to 10, this means that group A will be evaluated over 10 minutes, and group B will be evaluated over 100 minutes (which is 10 offset commits for each of them).

tl;dr - You should be able to do what you want with a single instance of Burrow. It should work regardless of how frequently your consumers commit offsets.

**Comment by aravindan-bk** [comment]
Thanks much for your clarification. So the Window slides whenever the "__consumer_offset" topic receives an offset commit message for a particular topic and partition. If that is the case, I have a follow up question, 
1.In Consumer Lag Evaluation Rules page - Example 2, the offset value remains the same for a few intervals. Is this because we always expect the Consumer to commit Offsets (even if they have not read new messages)?
We use SimpleConsumer Java client and we manage the Offset Commit ourselves. Today we commit offset only when we consume a message on Kafka. Wanted to know if we have to commit offsets regularly in order to use Burrow effectively.

**Comment by toddpalino** [comment]
When auto commit is enabled, that is what you can see. However, Burrow does not depend on this behavior. You do not need to commit offsets at a regular interval, you just need to understand that this means that the time window that you are evaluating lag over can change. So if you commit 10 offsets over the span of 5 minutes, and then the next 10 are over the span of 10 minutes, your evaluation window will slide from 5 minutes up to 10 minutes over the space of all those commits.

The one gotcha you can run into is that if you are committing offsets very quickly (in general, less than the broker-offsets config) you can run into a situation where the consumer group is incorrectly evaluated as being stale. This was identified in issue #24 and fixed with a new config option PR #25. This allows you to set a minimum time between consumer offset commits. Any commits that occur faster than this time from the previous commit are silently dropped.

**Comment by aravindan-bk** [comment]
Thanks for your response. I agree the auto-commit flag can help in case of High-Level consumers in Java which can take the ConsumerConfig object. Using simple consumer we manage offsets ourselves. Moreover issue #24 and PR #25 look great for consumers committing too frequently. Which will be useful for us too. 

In your example, the window slides at variable speed depending on the offsets committed by consumers. In scenario when none of the consumers commit offset for a topic (due to some error while processing the message downstream), the window won't slide at all and we at least have to wait till Rule 4 kicks in,  for us to identify that there is a problem. But when my consumers don't commit regularly, 
Tmaxidx (Time of last commit in window) - T0 (Time of first commit in window) cannot be used as a  reliable value in Rule 4. Because if my window size is 10 and if my 11th commit takes longer than Tmaxidx - T0, then I would get a false alarm. Also, if for some reason my 5th commit took un-reasonably long, I would end up waiting for a long time  

 Hence I think we might have to commit offsets at regular intervals (even when we don't consume messages). I hope this is an acceptable approach, because regular commit intervals is what burrow assumes as well looking at your response to #24.

**Comment by toddpalino** [comment]
Yes, I can see where irregular commits can result in some undesirable behavior like that. In general, I am dealing with consumers that are using auto commit, or are at least dealing with topics that are not slow, so they are committing fairly often regardless.

Is it possible for you to add logic to your app to have a minimum commit interval? That is, if there have been no messages processed in X milliseconds, recommit the last committed offset. This also can help to avoid staleness in the consumer coordinator in the Kafka broker as well, as it has a configuration to purge stale commits (though the time window is much longer - I think it defaults to a week).

It would be possible to add a defined max value to use for rule 4, but I'd like to avoid setting as many thresholds as possible here. Part of my intention is to get away from threshold-based monitoring, which tends to be a problem when you have many different types of consumers, and find ways to automatically detect problems.

**Comment by aravindan-bk** [comment]
Yes, the intention is to add logic to my application. Agree that threshold config for rule 4 might not be good in Burrow for the reason you mentioned. Closing this ticket. Thanks for your responses.
